# Peadiatric Acute Myeloid Leukemia (TARGET, 2018) - Exploratory Data Analysis

---

In [2]:
import warnings
warnings.filterwarnings("ignore", category=UserWarning)
warnings.filterwarnings("ignore", category=FutureWarning)

## Dataset Loading and Overview

### Pediatric Acute Myeloid Leukemia (TARGET, 2018)

This dataset contains clinical information from the **TARGET AML study** (2018), focused on paediatric acute myeloid leukaemia (AML). It includes:

- **1025 rows** (patients) and **75 columns** (features),
- Clinical details such as diagnosis age, survival, genetic mutations, and treatment response.

The goal is to load this dataset, ensure clean handling of missing values, and prepare it for exploratory analysis.

### Load and preview dataset

In [4]:
import pandas as pd

# Specify the file path
file_path = r'C:\Users\shali\Documents\L&D\GitHub Projects\Clinical Data Analysis\001_cBioPortal_Pediatric AML TARGET 2018\aml_target_2018_pub_clinical_data.xlsx'

# Load the Excel file, treating 'NA' as a missing value
clinical_data = pd.read_excel(file_path, na_values=['NA'])
clinical_data

Study ID        Patient ID            Sample ID  \
0     aml_target_2018_pub  TARGET-20-PABLDZ  TARGET-20-PABLDZ-04   
1     aml_target_2018_pub  TARGET-20-PABLDZ  TARGET-20-PABLDZ-09   
2     aml_target_2018_pub  TARGET-20-PADYIR  TARGET-20-PADYIR-03   
3     aml_target_2018_pub  TARGET-20-PADYIR  TARGET-20-PADYIR-04   
4     aml_target_2018_pub  TARGET-20-PADYIR  TARGET-20-PADYIR-09   
...                   ...               ...                  ...   
1020  aml_target_2018_pub  TARGET-21-PATKBK  TARGET-21-PATKBK-41   
1021  aml_target_2018_pub  TARGET-21-PATKKJ  TARGET-21-PATKKJ-09   
1022  aml_target_2018_pub  TARGET-21-PATKKJ  TARGET-21-PATKKJ-41   
1023  aml_target_2018_pub  TARGET-21-PATKWH  TARGET-21-PATKWH-09   
1024  aml_target_2018_pub  TARGET-21-PATKWH  TARGET-21-PATKWH-41   

      Diagnosis Age  Diagnosis Age (days)                Analysis Cohort  \
0               7.0                2455.0                      DISCOVERY   
1               7.0                2455.0                      DISCOVERY   
2               4.0                1159.0                      DISCOVERY   
3               4.0                1159.0                      DISCOVERY   
4               4.0                1159.0                      DISCOVERY   
...             ...                   ...                            ...   
1020            1.0                  31.0  INDUCTION FAILURE - DISCOVERY   
1021           24.0                8581.0  INDUCTION FAILURE - DISCOVERY   
1022           24.0                8581.0  INDUCTION FAILURE - DISCOVERY   
1023           17.0                6008.0  INDUCTION FAILURE - DISCOVERY   
1024           17.0                6008.0  INDUCTION FAILURE - DISCOVERY   

      Bone Marrow Leukemic Blast Percentage Bone Marrow Site of Relapse  \
0                                      93.0                         Yes   
1                                      93.0                         Yes   
2                                      29.0                         Yes   
3                                      29.0                         Yes   
4                                      29.0                         Yes   
...                                     ...                         ...   
1020                                   30.0                    Not done   
1021                                   65.0                    Not done   
1022                                   65.0                    Not done   
1023                                   20.0                         Yes   
1024                                   20.0                         Yes   

     Cancer Type    Cancer Type Detailed  ... Trisomy 8 t(10;11)(p11.2;q23)  \
0       Leukemia  Acute Myeloid Leukemia  ...   Unknown             Unknown   
1       Leukemia  Acute Myeloid Leukemia  ...   Unknown             Unknown   
2       Leukemia  Acute Myeloid Leukemia  ...   Unknown             Unknown   
3       Leukemia  Acute Myeloid Leukemia  ...   Unknown             Unknown   
4       Leukemia  Acute Myeloid Leukemia  ...   Unknown             Unknown   
...          ...                     ...  ...       ...                 ...   
1020    Leukemia  Acute Myeloid Leukemia  ...        No                  No   
1021    Leukemia  Acute Myeloid Leukemia  ...        No                  No   
1022    Leukemia  Acute Myeloid Leukemia  ...        No                  No   
1023    Leukemia  Acute Myeloid Leukemia  ...        No                  No   
1024    Leukemia  Acute Myeloid Leukemia  ...        No                  No   

     t(11:19)(q23:p13.1) t(3;5)(q25;q34) t(6;11)(q27;q23)   t(6;9)  t(8;21)  \
0                Unknown         Unknown          Unknown  Unknown  Unknown   
1                Unknown         Unknown          Unknown  Unknown  Unknown   
2                Unknown         Unknown          Unknown  Unknown  Unknown   
3                Unknown         Unknown          Unknown  Unknown  Unknown   
4                Unknown         Unknown  

### Dataset Overview

The first few rows of the dataset reveal:

- **Columns:** There are 75 columns covering various clinical, genetic, and demographic variables.
- **Missing Values:** Missing data is represented by "NA" (handled appropriately).
- **Data Types:** Columns like `Diagnosis Age`, `Bone Marrow Leukemic Blast Percentage`, and mutations appear to contain numeric and categorical data.


## Initial Data Exploration

### Column Names

In [14]:
# Get all column names
column_names = clinical_data.columns
print("Column Names:")
print(column_names)

Column Names:
Index(['Study ID', 'Patient ID', 'Sample ID', 'Diagnosis Age',
       'Diagnosis Age (days)', 'Analysis Cohort',
       'Bone Marrow Leukemic Blast Percentage', 'Bone Marrow Site of Relapse',
       'Cancer Type', 'Cancer Type Detailed', 'CBFT2A3-GLIS', 'CEBPA Mutation',
       'Chloroma', 'Chloroma Site of Relapse', 'C-Kit Mutation Exon 17',
       'C-Kit Mutation Exon 8', 'CNS Disease', 'CNS Site of Relapse',
       'CR Status at End of Course 1', 'CR Status at End of Course 2',
       'Cytogenetic Code (Other)', 'Cytogenetic Complexity',
       'Cytogenetic Relapse Site', 'Time To Event (days)', '5Q Deletion',
       '7Q Deletion', '9Q Deletion', 'Ethnicity Category', 'FAB',
       'First Event', 'FLT3-ITD Allelic Ratio.', 'FLT3-ITD Postive', 'FLT3-PM',
       'Fraction Genome Altered', 'inv(16)', 'ISCN', 'Minus-X', 'Minus-Y',
       'MLL', 'Monosomy 5', 'Monosomy 7', 'MRD at end of Course 1',
       'MRD at end of Course 2', 'MRD Percentage at end of Course 1',
      

### Count the number of columns with missing values

In [12]:
# Count the number of columns with missing values
num_columns_with_nan = len(columns_with_nan)
print(f"Number of columns with missing values: {num_columns_with_nan}")

Number of columns with missing values: 66


### Check for columns which have missing values

In [10]:
# Check for columns with missing values
missing_values = clinical_data.isnull().sum()

# Filter to show only columns with missing values
columns_with_nan = missing_values[missing_values > 0]
print("Columns with missing values:")
print(columns_with_nan)

Columns with missing values:
Diagnosis Age                             13
Diagnosis Age (days)                      13
Bone Marrow Leukemic Blast Percentage     72
Bone Marrow Site of Relapse               13
CBFT2A3-GLIS                             604
                                        ... 
t(6;9)                                    13
t(8;21)                                   13
t(9;11)(p22;q23)                          13
WBC                                       13
WT1 Mutation                              13
Length: 66, dtype: int64


In [16]:
# Sort missing values in descending order
columns_with_nan_sorted = columns_with_nan.sort_values(ascending=False)
print("Top 10 columns with missing values:")
print(columns_with_nan_sorted.head(10))

Top 10 columns with missing values:
Mutation Count                       875
TMB (nonsynonymous)                  875
FLT3-ITD Allelic Ratio.              855
Fraction Genome Altered              772
Cytogenetic Code (Other)             765
CBFT2A3-GLIS                         604
NUP98-NSD1                           604
Other MLL                            597
MRD Percentage at end of Course 2    323
MRD Percentage at end of Course 1    222
dtype: int64


### Shape of the data

In [267]:
clinical_data.shape

(1025, 75)

**Conclusions**

**Column Diversity**  
The dataset includes a wide range of features:  
- **Demographic data**: Ethnicity, age, and sex.  
- **Clinical outcomes**: Survival rates, remission statuses, and bone marrow assessments.  
- **Genetic and mutation data**: Cytogenetic codes, mutation statuses, and karyotype details.  

This diverse set of variables offers rich opportunities for exploratory data analysis, correlation studies, and predictive modelling.

**Missing Values**  
- A total of **66 columns** have missing values, which must be handled carefully to maintain data integrity.  
- Key variables with significant missing data include:  
   - **Mutation Count**: 875 missing  
   - **TMB (nonsynonymous)**: 855 missing  
   - **FLT3-ITD Allelic Ratio**: 755 missing  
   - **Bone Marrow Leukemic Blast Percentage**: 72 missing  
   - **MRD at End of Course 2**: 128 missing  

These variables are critical for downstream survival or mutation analyses, making appropriate imputation or cleaning strategies essential.

**Note:** Removing rows with missing values across all columns caused excessive data loss. To address this:
- Data cleaning will be performed selectively, 
- Rows will only be dropped for critical columns essential to the specific analysis.

This ensures we retain as much data as possible while maintaining data integrity.

**Dataset Structure**  
The dataset contains **1025 rows** and **75 columns**, providing a comprehensive clinical and genetic overview of paediatric AML patients.


## Preserving the Original Dataset

How can we ensure the integrity of the original dataset during data processing and analysis?

### Made a copy of the original DataFrame to preserve the original dataset

- To ensure the integrity of the dataset, we create a **copy** of the original DataFrame. This allows us to make modifications freely without altering the raw data, which is essential for reproducibility and traceability.

- **Why use `.copy()`?**  
Simply assigning the DataFrame (e.g., `clinical_data_copy = clinical_data`) creates a reference to the original data, so changes to the copy would affect the original.  
Using `.copy()` ensures an independent duplicate of the data.


In [33]:
# Create a copy of the original DataFrame
clinical_data_copy = clinical_data.copy()
clinical_data_copy.shape

(1025, 75)

**Conclusion:**

- A copy of the original DataFrame was successfully created to preserve the raw data. This ensures that any modifications made during analysis will not affect the original dataset.  
- Creating a backup safeguards against accidental data loss, irreversible changes, or errors during data processing—making it a best practice in data analysis workflows.


## Survival Analysis

### Cleaning and Inspecting Survival Data

**Questions Addressed:**

- How can we clean and prepare survival data (Overall Survival Days and Overall Survival Status) for analysis?
- Are there any missing or invalid values in these survival-related columns, and how should they be handled?

**Why Clean Survival Data?**

Survival analysis relies on clean and complete data for accurate results.  
Key variables include:  
- **Overall Survival Days**: Total days a patient survived from diagnosis.  
- **Overall Survival Status**: Binary outcome indicating whether the patient is **LIVING** or **DECEASED**.  

Ensuring these columns are free of missing or invalid values is a crucial first step in survival modelling.

We remove rows with missing `Overall Survival Days` because survival time is critical for analysis.  
Using `subset` ensures that only this column is checked, leaving other data untouched.


In [47]:
# Filter Data by Overall Survival Days
clinical_data_copy_os_days = clinical_data_copy.dropna(subset=['Overall Survival Days'])

# Check the shape of the cleaned DataFrame for overall survival days
print(f"Shape after cleaning 'Overall Survival Days': {clinical_data_copy_os_days.shape}")

# Display the first few rows if you want to inspect the cleaned data
print("First few rows of 'Overall Survival Days' and 'Overall Survival Status':")
print(clinical_data_copy_os_days[['Overall Survival Days', 'Overall Survival Status']].head(11))

# Check for missing values in these columns
print("\nMissing values in 'Overall Survival Days' and 'Overall Survival Status' after reloading:")
print(clinical_data_copy_os_days[['Overall Survival Days', 'Overall Survival Status']].isnull().sum())

Shape after cleaning 'Overall Survival Days': (1012, 75)
First few rows of 'Overall Survival Days' and 'Overall Survival Status':
    Overall Survival Days Overall Survival Status
0                   721.0                0:LIVING
1                   721.0                0:LIVING
2                   585.0              1:DECEASED
3                   585.0              1:DECEASED
4                   585.0              1:DECEASED
5                   536.0              1:DECEASED
6                   536.0              1:DECEASED
7                  1346.0                0:LIVING
8                   223.0              1:DECEASED
9                   596.0              1:DECEASED
10                  163.0              1:DECEASED

Missing values in 'Overall Survival Days' and 'Overall Survival Status' after reloading:
Overall Survival Days      0
Overall Survival Status    0
dtype: int64



- **Cleaning**: Rows with missing `Overall Survival Days` were successfully removed.  
- **Dataset Shape**: The cleaned dataset now contains **1012 rows** and **75 columns**.  
- **Validation**: There are no missing values in `Overall Survival Days` or `Overall Survival Status`.  

The cleaned data is now ready for survival analysis, such as Kaplan-Meier curves or Cox Proportional Hazards modelling.

### Preprocessing Survival Data for Analysis

In [49]:
# Extract numeric part from 'Overall Survival Status' by splitting at ':'
clinical_data_copy_os_days.loc[:, 'Overall Survival Status'] = clinical_data_copy_os_days['Overall Survival Status'].apply(
    lambda x: int(str(x).split(':')[0]) if pd.notnull(x) else x
)

# Convert 'Overall Survival Days' to numeric with coercion
clinical_data_copy_os_days.loc[:, 'Overall Survival Days'] = pd.to_numeric(clinical_data_copy_os_days['Overall Survival Days'], errors='coerce')

# Convert 'Overall Survival Status' to numeric with coercion
clinical_data_copy_os_days.loc[:, 'Overall Survival Status'] = pd.to_numeric(clinical_data_copy_os_days['Overall Survival Status'], errors='coerce')

# Drop rows with missing values in the survival columns
survival_data = clinical_data_copy_os_days.dropna(subset=['Overall Survival Days', 'Overall Survival Status'])

# Verify the cleaned data
print("First few rows after cleaning:")
print(survival_data[['Overall Survival Days', 'Overall Survival Status']].head())

print("\nRemaining missing values in survival columns after cleaning:")
print(survival_data[['Overall Survival Days', 'Overall Survival Status']].isnull().sum())

First few rows after cleaning:
   Overall Survival Days Overall Survival Status
0                  721.0                       0
1                  721.0                       0
2                  585.0                       1
3                  585.0                       1
4                  585.0                       1

Remaining missing values in survival columns after cleaning:
Overall Survival Days      0
Overall Survival Status    0
dtype: int64


**Conclusions:**

- Cleaning:

Rows with missing values in the Overall Survival Days and Overall Survival Status columns were removed, reducing the dataset size from 1025 rows to 1012 rows.
This ensures that the survival data is complete and reliable for analysis.

- Preprocessing:
- 
The Overall Survival Status column was cleaned by extracting the numeric survival status, ensuring it is binary and ready for analysis.The text after the colon (e.g., :LIVING or :DECEASED) is not necessary for survival analysis.
Both Overall Survival Days and Overall Survival Status were converted to numeric formats to maintain consistency for downstream analyses.
Missing or invalid values were dropped, leaving no remaining null entries in the survival columns.

### Install Lifelines Python package for plotting Kaplan Meier Curves

Why Lifelines?  

**Lifelines** is a Python library designed for survival analysis. It allows us to:  
- Plot Kaplan-Meier survival curves,  
- Perform statistical comparisons of survival data,  
- Fit and visualise models like Cox Proportional Hazards.  


In [51]:
pip install lifelines

Note: you may need to restart the kernel to use updated packages.


In [60]:
survival_data.columns

Index(['Study ID', 'Patient ID', 'Sample ID', 'Diagnosis Age',
       'Diagnosis Age (days)', 'Analysis Cohort',
       'Bone Marrow Leukemic Blast Percentage', 'Bone Marrow Site of Relapse',
       'Cancer Type', 'Cancer Type Detailed', 'CBFT2A3-GLIS', 'CEBPA Mutation',
       'Chloroma', 'Chloroma Site of Relapse', 'C-Kit Mutation Exon 17',
       'C-Kit Mutation Exon 8', 'CNS Disease', 'CNS Site of Relapse',
       'CR Status at End of Course 1', 'CR Status at End of Course 2',
       'Cytogenetic Code (Other)', 'Cytogenetic Complexity',
       'Cytogenetic Relapse Site', 'Time To Event (days)', '5Q Deletion',
       '7Q Deletion', '9Q Deletion', 'Ethnicity Category', 'FAB',
       'First Event', 'FLT3-ITD Allelic Ratio.', 'FLT3-ITD Postive', 'FLT3-PM',
       'Fraction Genome Altered', 'inv(16)', 'ISCN', 'Minus-X', 'Minus-Y',
       'MLL', 'Monosomy 5', 'Monosomy 7', 'MRD at end of Course 1',
       'MRD at end of Course 2', 'MRD Percentage at end of Course 1',
       'MRD Percenta